### ISSUE TITLE
Sending IPC messages to renderers interferes with node http module on windows

### ISSUE THREAD

**Issue by sentialx** [electron/electron#19368]
### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  *  6.0.0-beta.15
* **Operating System:**
  * Windows 10 1903
* **Last Known Working Electron version:**
  * 5.0.7

### Expected Behavior
Node's built-in should work all the time the app starts.

### Actual Behavior
I noticed when using NeDB in my Electron app in renderer process after the update to 6.0.0-beta.15, the databases just don't load (they load properly very rare). But when reloading my renderer process using `Ctrl + R` everything works fine. Also it's worth to mention that requests using `http` or `https` modules behave in the same way, they just sometimes don't work. The problem doesn't occur in Electron 5.0.7. The problem seems to be very random and for me it's hard to describe what causes it.

### To Reproduce
The best way to reproduce this is to build my [project](https://github.com/wexond/wexond). All instructions are in the README. Then try to add some bookmarks, close the app and restart it and see if they appear under the address bar. Also as I mentioned before, sometimes web requests doesn't work so you may not see the weather card. Then try to reload using `Ctrl + R` and you will see that everything works correctly.

**Comment by nornagon** [comment]
Can you please try to build a smaller test case, e.g. using [Electron Fiddle](https://github.com/electron/fiddle)?

**Comment by sentialx** [comment]
@nornagon How can I install there NeDB? I can't reproduce the bug without this module, which works correctly in 5.0.7.

**Comment by nornagon** [comment]
I'm confused why NeDB would be related as the issue is titled
> Node's built-in modules sometimes not working
and NeDB is not a built-in module. Is this an issue with NeDB or with a built-in Node module? If the former, then this should be filed against NeDB. If the latter, then it should be reproducible without any dependencies.

**Comment by sentialx** [comment]
I suspect that's a problem with `fs` module which NeDB is using, because the callback in `loadDatabase` is not being executed sometimes. This [issue ](https://github.com/louischatriot/nedb/issues/598) seems to be related, but the described behavior appears only in 6.0.0-beta.

**Comment by nornagon** [comment]
It would be great to narrow down a repro which only uses `fs` then!

**Comment by sentialx** [comment]
Okay, I will try to find what causes the issue. I'm closing this now.

**Comment by sentialx** [comment]
@nornagon After some investigation, I discovered that the issue is caused by WinAPI `GetForegroundWindow` call in `setInterval` each 50ms in my main process. In result it blocks Node's built-in modules such as `fs` or `http` in renderer process. Please let me know if you'd like me to create a repo which reproduces the following issue (I can't use Fiddle, since it doesn't support creating native modules).

**Comment by nornagon** [comment]
@sentialx a repro would be great. perhaps use [electron-quick-start](https://github.com/electron/electron-quick-start) as a basis for the repro?

**Comment by sentialx** [comment]
Okay, it turns out that's not a WinAPI call related issue. According to my observations, it's a `webContents.send` issue which blocks for example `http` from executing. Here's my repro: https://gist.github.com/sentialx/6e4eb680b3f6a84dd1ecb73caf181dfc

**Comment by nornagon** [comment]
@sentialx thanks for the repro case, this is very helpful. Just to make sure I'm seeing what you're seeing, could you describe (1) steps I can take, (2) what you expect to see, and (3) what you actually see?

**Comment by sentialx** [comment]
@nornagon 

(1) You can:
- Just open the app and see basically nothing
- Type in inspector `makeRequest` or `requestURL(your_url).then((d) => console.log(d.data));
- `Ctrl+R` to see the fetched content

(2) I expect to see response data from https://wexond.net in a `<div>`

(3) I see nothing

**Comment by nornagon** [comment]
Thanks. I can repro on Windows with Electron v6.0.0-beta.13. macOS seems fine.

**Comment by nornagon** [comment]
Possibly related: https://github.com/electron/electron/issues/19384

**Comment by nornagon** [comment]
Thanks a ton for your work on this @sentialx — the repro you created was super valuable in finding a fix!

**Comment by sentialx** [comment]
@nornagon No problem! Thanks for fixing it!

**Comment by sentialx** [comment]
@nornagon After upgrading to Electron 6.0.0 it just got even worse. The requests with `http` are fetching very rarily, however I can't reproduce it using Electron Fiddle though. The problem is really awkward, here's my code where it blocks the `http` module:

```typescript
windowManager.on('window-activated', (window: Window) => {
  this.appWindow.webContents.send('select-tab', window.id); // <-- When I remove this line, the problem doesn't occur.
});
```

The `window-activated` event is from my [node-window-manager](https://github.com/sentialx/node-window-manager) package. You can find it [here](https://github.com/sentialx/node-window-manager/blob/master/src/index.ts#L24). It's basically a `setInterval` with 50ms checking an active window. This event is fired only when a window has been activated.

![screen4](https://user-images.githubusercontent.com/11065386/62086632-458cbf80-b25f-11e9-92e0-9c7a61774ba1.gif)

**Comment by nornagon** [comment]
@sentialx i don't think this fix has been shipped yet.

**Comment by sentialx** [comment]
@nornagon Okay, after some testing it turns out I'm able to reproduce the problem. It appears it has problems when sending an IPC message to renderer and with registering an `ipcMain` event listener.

Here's my repro: https://gist.github.com/ccd0cd7f806aa00dfc5e85f3bb09cc39

**Comment by sentialx** [comment]
@nornagon Were you also able to reproduce this issue?

**Comment by MarshallOfSound** [comment]
I could reproduce this, going to re-open for some investigative work

**Comment by enriquepiatti** [comment]
I'm using NeDB and the callback is executed fine if I call loadDatabase from the Main process AND (this is important) before creating any window. After creating the window loadDatabase callback is not working anymore (I'm calling it always from Main process, never from Renderer).
Is this related with this issue or I should open a new one?

**Comment by ThatAnton** [comment]
I think this issue (or similar) still persists in regards to electron + nedb. My DBs act like they're loading but the files don't get compacted and there is no data found. Just an empty array for docs. I've tested the same code outside of electron and everything loads fine.

**Comment by whatbotisthis** [comment]
Was there ever a fix to this issue? I am running into a similar issue regarding NeDB like @enriquepiatti. However I only have a single Electron window and I'm loading my database inside of the renderer. I'm currently only experiencing this issue on Windows dev environment. The packaged app on Windows runs fine. And both dev and packaged work just fine on Mac OS.

**Comment by nornagon** [comment]
@whatbotisthis thanks for the info, would you mind opening a new issue? If possible, include a repro using [Electron Fiddle](/electron/fiddle) or [electron-quick-start](/electron/electron-quick-start).

**Comment by marcus13345** [comment]
@nornagon I'm having an extremely similar issue, that i believe to have the same root cause. Ive created an extremely simple test case [with electron fiddle](https://gist.github.com/marcus13345/affedc59a009210f18d3d396bb9cc425). as well as cross posted this test case on the [nedb repo](https://github.com/louischatriot/nedb/issues/649)

**Comment by nornagon** [comment]
@marcus13345 nedb is a huge module that does a lot of different things. Would be great to get a test case that just uses built-in Electron / Node, without any 3rd-party modules.

**Comment by marcus13345** [comment]
@nornagon I completely understand that, however, I am not familiar enough with the internal code of either electron, or nedb to be able to give adequate information. (hence why I've cross issued this). Hopefully, someone at nedb will be able to isolate the interaction. At the very least, this test case is what I have so far to describe the issue.

**Comment by nornagon** [comment]
Thanks. Hopefully someone familiar with NeDB can help us narrow this down.

**Comment by marcus13345** [comment]
@nornagon I've done some more digging, and this rabbit hole goes deep. I believe it to be a separate problem though, so I've made [a new issue](https://github.com/electron/electron/issues/25314)

**Comment by whatbotisthis** [comment]
@nornagon So I figured out what was causing my issue but was unable to reproduce it in a fiddle. While trying to track this issue down I commented out a call to openDevTools() which I was making right after calling mainWindow.loadURL(). Suddenly NeDB started working on Windows dev for me. This made sense as my production build on Windows (which does not have devtools) worked fine.

I have now moved my openDevTools() call inside of the ready-to-show event listener and everything is working fine. I'm wondering if calling openDevTools() too soon does something similar to @enriquepiatti's issue involving multiple windows. My app only has a single window and load NeDB inside the renderer.

Sorry I wasn't able to reproduce in a fiddle but hopefully this provides some insight to anyone capable of fixing this issue. Sadly this is about where my abilities stop.
